### ISSUE TITLE
Idea: Whole assembly decompilation report

### ISSUE THREAD

**Issue by tamlin-mike** [icsharpcode/ILSpy#1802]
Looking through some decompiled code, I came across a `double` literal `0.10000000149011612`. That's obviously a rounding error, and should be `0.1`. But that got me thinking.

Remember the problem with MathF references being generated even that it was only a netstandard2.0 assembly, or when float helpers turned a 255.f/smth (or if it was 256.f/smth) into something like 679/431 (completely made up numbers)?

When ILSpy is helpful, but have the potential of making a bad call, it could be good to have all such decisions in a "ProjectDecompilationReport" of sorts (a saved file, alongside the csproj). Not only for math-related issues (that was just what got me thinking of this), but also a place to inform the user of problems currently only reported in a comment in the decompiled src, a comment that can be quite well hidden unless you grep - and you know what to grep for.

This file could very well be chatty, since it'd be a one-off thing. For math tranforms it could f.ex. emit not only the r4 or r8 text representations, but maybe even hex, in addition to what tranform it performed. It could tell you that it `OR`ed a bunch of `Flags` together, or when some enumeratees had the same value so it's not certain of the actual identifier intended (I've seen that in the live, where f.ex. multiple enumeratees have the value zero). Where it made a decision to emit something as hex instead of decimal (or vice versa), and last but surely not least a section listing problems it encountered.
(these are just some sections from the top of my head, I'm sure there could be other of interest)

Not only could it be helpful to be able, at a glance, keep a watchful eye on what was done, but it could possibly also help in finding some otherwise elusive problems.

I consider this a very low priority idea, but let it sink in for a while. Sleep on it. It struck me less than an hour ago, but it's already starting to grow on me. I believe it might have some merit.

Comments?

**Comment by dgrunwald** [comment]
`0.10000000149011612` is the same as `(double)(float)0.1`, which is not the same as `0.1`. The original code must have had some weird casts to arrive at this value.

For decompilation warnings (when the IL cannot be expressed in C#) I agree, it could be helpful to collect those into a report.
Floating-point math is not a part of that: ILSpy should always produce a number that results in a bit-identical value when recompiled.

Generally the goal for ILSpy is to produce code that, when recompiled, will have the same runtime behavior as the original. It's not a goal to reproduce the original C# code. It's not a goal to produce code that re-compiles to the same IL.
There's a whole bunch of constructs where we can't know what the original C# code was. Every bit of syntax sugar is ambiguous between the high-level form (e.g. `foreach`) and the low-level form (`try-finally` + `while(e.MoveNext()`) and maybe some forms in between (`using` + `while (e.MoveNext())`). Every instance call on a value type involves an IL-level local, but whether that was user-declared or compiler-generated is unknown, so ILSpy constantly takes guesses. For every `if`, we have to guess which branch came first in source order -- the compiler may have negated the condition and swapped the two branches (this happens frequently if one of the branches is just a `return;` or `break;`). It would be a lot of work to annotate all this in the decompiler for a report, and the report would be so verbose that it would be useless.

**Comment by tamlin-mike** [comment]
> 0.10000000149011612 is the same as (double)(float)0.1

You are right. The literal probably started its life as a `const float`, then compiler-promoted to be used used in a `double` context.

> [...] It would be a lot of work to annotate all this in the decompiler for a report, and the report would be so verbose that it would be useless.

You may be right. I was thinking of a limited subset of transformations, when in reality there are so many that could potentially also be considered, that it's simply not feasible - the report size could dwarf the decompiled output.

Leaving this issue open though, as it seems the part about collecting decompilation warnings might get traction.
